transaction/LockManager.cpp

```diff
@@ -0,0 +1,235 @@
+                                                            &detector_status_,
+                                                            &victim_result_)),
+      incoming_requests_(*incoming_requests),
+      permitted_requests_(*permitted_requests),
+      inner_pending_requests_() {
+}
+
+LockManager::~LockManager() {
+  deadlock_detector_->join();
+}
+
+void LockManager::run() {
+  deadlock_detector_->start();
+
+  constexpr std::uint64_t max_try_incoming = 10000;
+  constexpr std::uint64_t max_try_inner = 6000;
+
+  while (true) {
+    for (std::uint64_t tries = 0; tries < max_try_incoming; ++tries) {
+      if (!incoming_requests_.empty()) {
+        LockRequest request = incoming_requests_.popOne();
+        if (request.getRequestType() == RequestType::kReleaseLocks) {
+          bool result = releaseAllLocks(request.getTransactionId());
+          if (!result) {
+            LOG(FATAL) << "Unexpected condition occured.";
+          }
+        } else {
+          bool result = acquireLock(request.getTransactionId(),
+                                    request.getResourceId(),
+                                    request.getAccessMode());
+          if (!result) {
+            LOG(INFO) << "Transaction "
+                         + std::to_string(request.getTransactionId())
+                         + " is waiting " + request.getResourceId().toString();
+
+            inner_pending_requests_.push(request);
+          } else {
+            LOG(INFO) << "Transaction "
+                         + std::to_string(request.getTransactionId())
+                         + " acquired " + request.getResourceId().toString();
+
+            permitted_requests_.push(request);
+          }
+        }
+      }
+    }
+    for (std::uint64_t tries = 0; tries < max_try_inner; ++tries) {
+      if (!inner_pending_requests_.empty()) {
+        LockRequest request = inner_pending_requests_.front();
+
+        bool result = acquireLock(request.getTransactionId(),
+                                  request.getResourceId(),
+                                  request.getAccessMode());
+        if (result) {
+          inner_pending_requests_.pop();
+          permitted_requests_.push(request);
+        }
+      }
+    }
+
+    killVictims();
+  }
+}
+
+bool LockManager::acquireLock(const transaction_id tid,
+                              const ResourceId &rid,
+                              const AccessMode access_mode) {
+  std::stack<std::pair<ResourceId, AccessMode>> stack;
+  ResourceId current_rid = rid;
+  AccessMode current_access_mode = access_mode;
+  stack.push(std::make_pair(current_rid, current_access_mode));
+
+  while (current_rid.hasParent()) {
+    current_rid = current_rid.getParentResourceId();
+    current_access_mode =
+      (current_access_mode.isShareLock() ||
+       current_access_mode.isIntentionShareLock())
+      ? AccessMode(AccessModeType::kIsLock)
+      : AccessMode(AccessModeType::kIxLock);
```

Comment:
According to the code style for breaking a line, we need two more whitespace indentations:

https://github.com/pivotalsoftware/quickstep/wiki/Quickstep-Style-Guide